File: toy/sns/src/main/java/me/kzv/sns/exception/GlobalControllerAdvice.java
package me.kzv.sns.exception;

import lombok.extern.slf4j.Slf4j;
import me.kzv.sns.controller.response.Response;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice
public class GlobalControllerAdvice {

    @ExceptionHandler(SnsApplicationException.class)
    public ResponseEntity<?> applicationHandler(SnsApplicationException e) {
        log.error("Error occurs {}", e.toString());
        return ResponseEntity.status(e.getErrorCode().getStatus())
                .body(Response.error(e.getErrorCode().name()));
    }
}


/**
 * name()
 * 열거 객체의 문자열 리턴
 *
 * Season season = Season.SPRING;
 * String name = season.name(); // "SPRING"
 */

File: toy/simple-sns/src/main/kotlin/me/kzv/simplesns/exception/GlobalControllerAdvice.kt
package me.kzv.simplesns.exception

import me.kzv.simplesns.controller.response.Response
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.lang.RuntimeException

@RestControllerAdvice
class GlobalControllerAdvice {

    @ExceptionHandler(SimpleSnsException::class)
    fun applicationHandler(e: SimpleSnsException): ResponseEntity<*> {
        return ResponseEntity.status(e.errorCode.status!!).body(Response.error(e.errorCode.name))
    }

    @ExceptionHandler(RuntimeException::class)
    fun applicationHandler(e: RuntimeException): ResponseEntity<*> {
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Response.error(ErrorCode.SERVER_ERROR.name))
    }
}